scripts/west_commands/run_common.py

```diff
@@ -165,15 +175,56 @@ def do_run_common(command, user_args, user_runner_args, domains=None):
             # Get the user specified domains.
             domains = load_domains(build_dir).get_domains(user_args.domain)
 
-    if len(domains) > 1 and len(user_runner_args) > 0:
-        log.wrn("Specifying runner options for multiple domains is experimental.\n"
-                "If problems are experienced, please specify a single domain "
-                "using '--domain <domain>'")
+    if len(domains) > 1:
+        if len(user_runner_args) > 0:
+            log.wrn("Specifying runner options for multiple domains is experimental.\n"
+                    "If problems are experienced, please specify a single domain "
+                    "using '--domain <domain>'")
+
+        # Process all domains to load board names and populate flash runner
+        # parameters.
+        for d in domains:
+            if d.build_dir is None:
+                build_dir = get_build_dir(user_args)
+            else:
+                build_dir = d.build_dir
+
+            cache = load_cmake_cache(build_dir, user_args)
+            board = cache['CACHED_BOARD']
+
+            # Load board flash runner configuration (if it exists) and store
+            # single-use commands in a dictionary so that they get executed
+            # once per unique board name.
+            try:
+                processed_boards[cache['BOARD_DIR']]
+            except KeyError:
+                board_runner_test_file = Path(cache['BOARD_DIR']) / 'flash_runner.yml'
+
+                try:
+                    with open(board_runner_test_file, 'r') as f:
+                        board_runner_test_yaml = yaml.safe_load(f.read())
+                    processed_boards[cache['BOARD_DIR']] = True
```

Comment:
Doing this at the board level feels like it's going to be pretty duplicative. Is there any way we can do this at the SoC level instead?

Since we have a build directory and therefore a devicetree, we can look up the `compatible` property in the `/soc` node from edt.pickle. We could compare that to runner-specific information for each SoC that we get using cmake target properties that we set in soc/.../CMakeLists.txt files which say what to do instead of having to repeat this for every board. We have a `runners_yaml_props_target` already in place for this that seems like it would potentially be usable.

My concern here is maintainability and DRY.

Reply:
The idea would be to do this at the SoC level, but that would be a part of boards/soc v2

Reply:
> The idea would be to do this at the SoC level, but that would be a part of boards/soc v2

But then in the meantime we are forced to carry a new config file format that might start appearing in out of tree boards, and thus needs to be maintained forever? I really would like to avoid adding something to the board file system conventions that we know we are going to throw away if possible.

Let me try another suggestion now that I've had another day to think this over: why not add new class methods to the ZephyrBinaryRunner API that let you query the information you need?

Something like this:

```diff
diff --git a/scripts/west_commands/runners/core.py b/scripts/west_commands/runners/core.py
index 7ed95d5bd1..656feb6627 100644
--- a/scripts/west_commands/runners/core.py
+++ b/scripts/west_commands/runners/core.py
@@ -523,6 +523,27 @@ class ZephyrBinaryRunner(abc.ABC):
     def do_add_parser(cls, parser):
         '''Hook for adding runner-specific options.'''
 
+    @classmethod
+    def one_time_options(cls) -> Set[str]:
+        '''Options that should be given only once when using the same
+        runner multiple times to flash different binaries onto the
+        same target.
+
+        Individual runners can extend the behavior of this method,
+        but should make sure to call the superclass first, like this:
+
+        @classmethod
+        def one_time_options(cls):
+            ret = cls.one_time_options()
+            ret.add('--my-runner-specific-option')
+            return ret
+        '''
+        ret = set()
+        caps = cls.capabilities()
+        if caps.erase:
+            ret.add('erase')
+        return ret
+
     @classmethod
     def create(cls, cfg: RunnerConfig,
                args: argparse.Namespace) -> 'ZephyrBinaryRunner':
```

And similarly for any other types of options that might be needed.

Then you could localize this information in the runners themselves instead of adding it to the boards, and query it from run_common.py appropriately.

Reply:
> Let me try another suggestion now that I've had another day to think this over: why not add new class methods to the ZephyrBinaryRunner API that let you query the information you need?

cc @trond-snekvik @bencefr

Reply:
The problem is that grouping is needed - 2 cases in point, nrf9160dk and nrf5340dk:

- nRF5340dk has application core and network core, these core are independent, but there is some linkage, so for the application core, doing --recover or --erase or --reset will only do that operation on that core, doing --erase or --reset on the network core will only do that operation on the network core, but doing --recover on the network core will affect **both** cores.
- nRF5340dk cores are linked, so the network core does not actually need to be reset **if** an application core image is flashed, because the reset of that will reset the network core
- nRF9160dk has an nrf52840 and nrf9160 - both completely separate MCUs, a reset, erase or recover of one has no effect on the other core
- There could be a third party board, similar to the nrf9160dk, but with a different setup, one of the MCUs could be in control of the power of the other with e.g. a FET, so it can power it down entirely to save power, so from a soc v2 perspective, this is now a different/incompatible design to the nrf9160dk

So there is always going to be a need to set these options, ideally at the SoC level, but also at the board level because some boards might want to override SoC configuration/settings